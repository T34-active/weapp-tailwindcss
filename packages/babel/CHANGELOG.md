# @weapp-tailwindcss/babel

## 0.0.1

### Patch Changes

- [`e0c37f5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e0c37f5f546b143341a75701a1907f876df38fa9) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore(deps): upgrade

## 0.0.1-alpha.0

### Patch Changes

- [`e0c37f5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e0c37f5f546b143341a75701a1907f876df38fa9) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore(deps): upgrade
