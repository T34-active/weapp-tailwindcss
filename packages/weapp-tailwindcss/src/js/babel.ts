import type { ParseError, ParseResult } from '@babel/parser'
import type { NodePath, TraverseOptions } from '@babel/traverse'
import type { File, Node } from '@babel/types'
import type { IJsHandlerOptions, JsHandlerResult } from '../types'
import { regExpTest } from '@/utils'
import { jsStringEscape } from '@ast-core/escape'
import MagicString from 'magic-string'
import { parse, traverse } from '../babel'
import { replaceHandleValue } from './handlers'
import { JsTokenUpdater } from './JsTokenUpdater'

export function isEvalPath(p: NodePath<Node>) {
  if (p.isCallExpression()) {
    const calleePath = p.get('callee')
    return calleePath.isIdentifier() && calleePath.node.name === 'eval'
  }
  return false
}

// function replaceStringLiteralsInBinaryExpr(binaryNode: BinaryExpression, ms: MagicString) {
//   if (t.isStringLiteral(binaryNode.left)) {
//     const leftStringLiteral = binaryNode.left
//     leftStringLiteral.value = 'bg-[#654321]' // 替换
//     ms.update(leftStringLiteral.start, leftStringLiteral.end, `'${leftStringLiteral.value}'`)
//   }
//   else if (t.isBinaryExpression(binaryNode.left)) {
//     replaceStringLiteralsInBinaryExpr(binaryNode.left, ms) // 递归
//   }

//   if (t.isStringLiteral(binaryNode.right)) {
//     const rightStringLiteral = binaryNode.right
//     rightStringLiteral.value = 'text-[#222222]' // 替换
//     ms.update(rightStringLiteral.start, rightStringLiteral.end, `'${rightStringLiteral.value}'`)
//   }
//   else if (t.isBinaryExpression(binaryNode.right)) {
//     replaceStringLiteralsInBinaryExpr(binaryNode.right, ms) // 递归
//   }
// }

const ignoreFlagMap = new WeakMap()

export function jsHandler(rawSource: string, options: IJsHandlerOptions): JsHandlerResult {
  const ms = new MagicString(rawSource)

  let ast: ParseResult<File>
  try {
    ast = parse(rawSource, options.babelParserOptions)
  }
  catch (error) {
    return {
      code: rawSource,
      error: error as ParseError,
    } as JsHandlerResult
  }
  let ignoreFlag = false
  const jsTokenUpdater = new JsTokenUpdater({ ignoreCallExpressionIdentifiers: options.ignoreCallExpressionIdentifiers })

  const traverseOptions: TraverseOptions<Node> = {
    StringLiteral: {
      enter(p) {
        if (ignoreFlag) {
          return
        }
        if (isEvalPath(p.parentPath)) {
          return
        }

        const n = p.node

        jsTokenUpdater.addToken(
          replaceHandleValue(
            n.value,
            n,
            {
              ...options,
              needEscaped: options.needEscaped ?? true,
            },
            1,
          ),
        )
      },
    },
    TemplateElement: {
      enter(p) {
        if (ignoreFlag) {
          return
        }
        const pp = p.parentPath
        if (pp.isTemplateLiteral()) {
          const ppp = pp.parentPath
          if (isEvalPath(ppp)) {
            return
          }
          if (ppp.isTaggedTemplateExpression()) {
            const tagPath = ppp.get('tag')
            if (
              (tagPath.isIdentifier()
                && regExpTest(options.ignoreTaggedTemplateExpressionIdentifiers ?? [], tagPath.node.name)
              )
            ) {
              return
            }
          }
        }
        const n = p.node
        jsTokenUpdater.addToken(
          replaceHandleValue(
            n.value.raw,
            n,
            {
              ...options,
              needEscaped: false,
            },
            0,
          ),
        )
      },
    },
    CallExpression: {
      enter(p) {
        if (isEvalPath(p)) {
          p.traverse({
            StringLiteral: {
              enter(s) {
                // ___CSS_LOADER_EXPORT___
                const res = jsHandler(s.node.value, {
                  ...options,
                  needEscaped: false,
                  generateMap: false,
                })
                if (res.code) {
                  const node = s.node
                  if (typeof node.start === 'number' && typeof node.end === 'number') {
                    const start = node.start + 1
                    const end = node.end - 1
                    if (start < end && s.node.value !== res.code) {
                      jsTokenUpdater.addToken(
                        {
                          start,
                          end,
                          value: jsStringEscape(res.code),
                        },
                      )
                    }
                  }
                }
              },
            },
            TemplateElement: {
              enter(s) {
                const res = jsHandler(s.node.value.raw, {
                  ...options,
                  generateMap: false,
                })
                if (res.code) {
                  const node = s.node
                  if (typeof node.start === 'number' && typeof node.end === 'number') {
                    const start = node.start
                    const end = node.end
                    if (start < end && s.node.value.raw !== res.code) {
                      jsTokenUpdater.addToken(
                        {
                          start,
                          end,
                          value: res.code,
                        },
                      )
                    }
                  }
                }
              },
            },
          })
          return
        }
        const callee = p
          .get('callee')
        if (
          callee
            .isIdentifier()
            && (
              regExpTest(
                options
                  .ignoreCallExpressionIdentifiers ?? [],
                callee
                  .node
                  .name,
              )

            )
        ) {
          ignoreFlag = true
          ignoreFlagMap.set(p, true)
        }
      },
      exit(p) {
        const flag = ignoreFlagMap.get(p)
        if (flag) {
          ignoreFlag = false
          ignoreFlagMap.delete(p)
        }
      },
    },
  }

  traverse(ast, traverseOptions)
  jsTokenUpdater.updateMagicString(ms)

  return {
    code: ms.toString(),
    get map() {
      return ms.generateMap()
    },
  }
}
