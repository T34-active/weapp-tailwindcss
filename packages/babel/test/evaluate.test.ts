import babel from '@babel/core'
import fs from 'fs-extra'
import path from 'pathe'

function getCase(name: string) {
  return fs.readFileSync(path.resolve(import.meta.dirname, './fixtures/evaluate', name), 'utf8')
}

const plugin: babel.PluginItem = {
  visitor: {
    BinaryExpression(path) {
      // 尝试计算二元表达式的常量值
      const evaluated = path.evaluate()

      // 如果可以计算，且计算结果是确定的常量
      if (evaluated.confident) {
        // 直接用计算后的常量值替换表达式
        path.replaceWithSourceString(evaluated.value.toString())
      }
    },

    CallExpression(path) {
      const callee = path.get('callee')

      // 确保是 Math.pow 函数调用
      const isMathPowCall
        = callee.isMemberExpression()
          && callee.get('object').isIdentifier({ name: 'Math' })
          && callee.get('property').isIdentifier({ name: 'pow' })

      if (isMathPowCall) {
        // 计算 Math.pow 的常量值
        const evaluated = path.evaluate()

        if (evaluated.confident) {
          // 如果能够确定计算结果，替换为常量值
          path.replaceWithSourceString(evaluated.value.toString())
        }
      }
    },
    ArrayExpression(path) {
      // 如果数组中的所有元素都是常量
      const evaluated = path.evaluate()

      if (evaluated.confident) {
        // 直接替换为计算后的值
        path.replaceWithSourceString(JSON.stringify(evaluated.value))
      }
    },
    ObjectExpression(path) {
      const evaluated = path.evaluate()

      if (evaluated.confident) {
        // 直接替换为计算后的对象
        path.replaceWithSourceString(JSON.stringify(evaluated.value))
      }
    },
    IfStatement(path) {
      // 计算条件表达式
      const evaluated = path.get('test').evaluate()

      if (evaluated.confident) {
        // 如果条件是常量，直接折叠成 `true` 或 `false`
        path.get('test').replaceWithSourceString(evaluated.value.toString())
      }
    },
  },
}
describe('evaluate', () => {
  it('0.js', () => {
    const res = babel.transform(
      getCase('0.js'),
      {
        plugins: [
          plugin,
        ],
      },
    )
    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('1.js', () => {
    const res = babel.transform(
      getCase('1.js'),
      {
        plugins: [
          plugin,
        ],
      },
    )
    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })
})
