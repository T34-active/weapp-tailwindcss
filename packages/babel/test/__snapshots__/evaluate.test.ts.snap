// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`evaluate > 0.js 1`] = `
"const a = 21;
const result = 8;
if (true) {
  console.log('This is true');
}"
`;

exports[`evaluate > 1.js 1`] = `
"const a = b + 1 + '1';
const result = 8;"
`;
